lib/rpc/src/tx_handler.rs

```diff
@@ -67,6 +73,60 @@ impl<Mempool: L2TransactionPool> TxHandler<Mempool> {
 
         Ok(hash)
     }
+
+    pub async fn send_raw_transaction_sync_impl(
+        &self,
+        bytes: Bytes,
+        max_wait_ms: Option<U256>,
+    ) -> Result<ZkTransactionReceipt, EthSendRawTransactionSyncError> {
+        let timeout_duration = if let Some(timeout_ms) = max_wait_ms {
+            match timeout_ms.try_into() {
+                Ok(timeout_u64) => {
+                    let requested_timeout = Duration::from_millis(timeout_u64);
+                    if requested_timeout > self.config.send_raw_transaction_sync_max_timeout {
+                        // Per EIP-7966 MUST use default timeout if user provided timeout is invalid
+                        self.config.send_raw_transaction_sync_timeout
+                    } else {
+                        requested_timeout
+                    }
+                }
+                Err(_) => {
+                    // Per EIP-7966 MUST use default timeout if user provided timeout is invalid
+                    self.config.send_raw_transaction_sync_timeout
+                }
+            }
+        } else {
+            self.config.send_raw_transaction_sync_timeout
+        };
+
+        // Create block subscription
+        let mut block_rx = self.storage.block_subscriptions().subscribe_to_blocks();
+
+        let tx_hash = self.send_raw_transaction_impl(bytes).await?;
+
+        // Wait for the transaction to appear in a block or timeout
+        tokio::time::timeout(timeout_duration, async {
+            loop {
+                // Wait for the next block notification
+                let Ok(block) = block_rx.recv().await else {
+                    // Channel closed, this shouldn't happen in normal operation
+                    tracing::warn!("block subscription closed while waiting for tx receipt");
+                    return Err(EthSendRawTransactionSyncError::Timeout(timeout_duration));
+                };
+
+                if let Some(stored_tx) = block.transactions.get(&tx_hash) {
+                    return Ok(build_api_receipt(
+                        tx_hash,
+                        stored_tx.receipt.clone(),
+                        &stored_tx.tx,
+                        &stored_tx.meta,
+                    ));
+                }
+            }
+        })
+        .await
+        .map_err(|_| EthSendRawTransactionSyncError::Timeout(timeout_duration))?
```

Comment:
I refactored your code a bit - added proper error types, moved logic to `TxHandler` and made so that block notifications are reused for receipts. Also added tests. LMK if you disagree with anything. If not we can merge I think.